### PR TITLE
sql: gate tenant's multi-region abstraction usage behind cluster setting

### PR DESCRIPTION
**Pull request by arulajmani**
This patch introduces a tenant read-only cluster setting called
`sql.multi_region.allow_abstractions_for_secondary_tenants.enabled`
which allows the operator to control if secondary tenants can make use
of multi-region abstractions. It defaults to false.

This setting is checked against when creating new MR databases or
altering existing ones to make use of MR features (by setting the
primary region). It has nothing to do with MR databases that may have
been created previously which could happen if this setting was ever
flipped to true by the operator.

This setting doesn't effect the system tenant. It also takes precedence
over `sql.defaults.primary_region`, which is tenant writeable.

Release note (sql change): introduces new cluster setting which allows
the operator to control if a secondary tenant can make use of MR
abstractions. The setting is called
`sql.multi_region.allow_abstractions_for_secondary_tenants.enabled`.

### CODE REVIEW

File: pkg/sql/logictest/logic.go [hunk 1]
```diff
@@ -2115,6 +2150,19 @@ type tenantClusterSettingOverrideOpt interface {
 	apply(*tenantClusterSettingOverrideArgs)
 }
 
+// tenantClusterSettingOverrideMultiTenantMultiRegionAbstractionsAllowed
+// corresponds to the allow-multi-region-abstractions-for-secondary-tenants
+// directive.
+type tenantClusterSettingOverrideMultiTenantMultiRegionAbstractionsAllowed struct{}
+
+var _ tenantClusterSettingOverrideOpt = &tenantClusterSettingOverrideMultiTenantMultiRegionAbstractionsAllowed{}
```
Comment (irfansharif) [inline pkg/sql/logictest/logic.go:2158]: That's a lot of words 😛

File: pkg/sql/descriptor.go [hunk 2]
```diff
@@ -361,6 +361,23 @@ var DefaultPrimaryRegion = settings.RegisterStringSetting(
 	"",
 ).WithPublic()
 
+const secondaryTenantsMultiRegionAbstractionsEnabledSettingName = "sql.multi_region.abstractions.enabled"
```
Comment (irfansharif) [inline pkg/sql/descriptor.go:364]: Include "tenant" in the setting name somewhere?

**Review by irfansharif — APPROVED** [review]

**Review by arulajmani — COMMENTED** [review]
> Dismissed @irfansharif from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79536)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @irfansharif)

___
*[pkg/sql/descriptor.go, line 364 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79536#-N--N5qypGN6XqCiptQj-r1-364:-N-3gQibEKqK4q3r-h_q:b-m46eme) ([raw file](https://github.com/cockroachdb/cockroach/blob/740529335479944e418a8a82b6993df67cecee80/pkg/sql/descriptor.go#L364)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Include "tenant" in the setting name somewhere?
</blockquote></details>

Good point, done. 
___
*[pkg/sql/logictest/logic.go, line 2158 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/79536#-N--N5qypGN6XqCiptQk-r1-2158:-N-3gVnf39NJ7LIphSdV:b-iot6t7) ([raw file](https://github.com/cockroachdb/cockroach/blob/740529335479944e418a8a82b6993df67cecee80/pkg/sql/logictest/logic.go#L2158)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

That's a lot of words 😛 
</blockquote></details>

:rocket: :rocket:

**Review by ajstorm — COMMENTED** [review]
Skimmed this and it seems reasonable to me.  One question: do we want to prevent restoring MR-enabled databases to clusters with this setting set to false?  If we don't it seems to be a pretty easy back-door to walk through.

> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79536#-:-N-433ns9TXKb8IJOJlJ:btjioer)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, and @irfansharif)

___
*[pkg/sql/descriptor.go, line 403 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/79536#-N-42Ca98tB-uChIbSIi:-N-42Ca98tB-uChIbSIj:blwu980) ([raw file](https://github.com/cockroachdb/cockroach/blob/d56af6d7ff8c740ac6c49249cb9d28d059e407df/pkg/sql/descriptor.go#L403)):*
> ```Go
> 		return nil, errors.WithHint(pgerror.Newf(
> 			pgcode.InvalidDatabaseDefinition,
> 			"setting %s disallows use of multi-region abstractions",
> ```

Nit: If this is going to be seen by Serverless users, is this the right error here?  Maybe something like "multi-region abstractions have been disabled on this cluster".

**Review by arulajmani — COMMENTED** [review]
> Skimmed this and it seems reasonable to me. One question: do we want to prevent restoring MR-enabled databases to clusters with this setting set to false? If we don't it seems to be a pretty easy back-door to walk through.

That's a good catch. I tacked on a second commit here which checks the value of this cluster setting when a secondary tenant tries to restore a MR database. @adityamaru would you mind giving the second commit a quick look? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79536#-:-N-OoQe01qg3hQWStKvq:b-zh90xz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, @irfansharif, and @shermanCRL)

**Review by adityamaru — APPROVED** [review]

**Review by arulajmani — COMMENTED** [review]
TFTRs! 

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/79536#-:-N-jQSw81QTV5xjy5jAc:b-oe2moi)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @arulajmani, @irfansharif, and @shermanCRL)

___
*[pkg/ccl/backupccl/restore_planning.go, line 1957 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/79536#-N-OoJTuJ7tj4tD1sfRI-r4-1957:-N-jQQqzAusdTXLtRpmA:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/acfd5ac62c87ac59f35fab26bf7b6238a875c69d/pkg/ccl/backupccl/restore_planning.go#L1957)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

nit: comment is truncated
</blockquote></details>

Done.

**Pull request merged**